session-helper/background-monitor.c

```diff
@@ -0,0 +1,426 @@
+                                                   NULL,
+                                                   &error))
+    {
+      g_dbus_error_strip_remote_error (error);
+      g_debug ("No background permissions found: %s", error->message);
+      return NULL;
+    }
+
+  return g_steal_pointer (&out_perms);
+}
+
+static Permission
+get_permission (const char *app_id,
+                GVariant   *perms)
+{
+  const char **permissions;
+
+  if (!g_variant_lookup (perms, app_id, "^a&s", &permissions))
+    {
+      g_debug ("No permissions stored for: app %s", app_id);
+
+      return UNSET;
+    }
+  else if (g_strv_length ((char **)permissions) != 1)
+    {
+      g_autofree char *a = g_strjoinv (" ", (char **)permissions);
+      g_warning ("Wrong permission format, ignoring (%s)", a);
+      return UNSET;
+    }
+  g_debug ("permission store: app %s -> %s", app_id, permissions[0]);
+
+  if (strcmp (permissions[0], "yes") == 0)
+    return YES;
+  else if (strcmp (permissions[0], "no") == 0)
+    return NO;
+  else if (strcmp (permissions[0], "ask") == 0)
+    return ASK;
+  else
+    {
+      g_autofree char *a = g_strjoinv (" ", (char **)permissions);
+      g_warning ("Wrong permission format, ignoring (%s)", a);
+    }
+
+  return UNSET;
+}
+
+static void
+set_permission (const char *app_id,
+                Permission  permission)
+{
+  g_autoptr(GError) error = NULL;
+  const char *permissions[2];
+
+  if (permission == ASK)
+    permissions[0] = "ask";
+  else if (permission == YES)
+    permissions[0] = "yes";
+  else if (permission == NO)
+    permissions[0] = "no";
+  else
+    {
+      g_warning ("Wrong permission format, ignoring");
+      return;
+    }
+  permissions[1] = NULL;
+
+  if (!xdp_dbus_permission_store_call_set_permission_sync (get_permission_store (),
+                                                           PERMISSION_TABLE,
+                                                           TRUE,
+                                                           PERMISSION_ID,
+                                                           app_id,
+                                                           (const char * const*)permissions,
+                                                           NULL,
+                                                           &error))
+    {
+      g_dbus_error_strip_remote_error (error);
+      g_warning ("Error updating permission store: %s", error->message);
+    }
+}
+
+static gboolean
+flatpak_instance_has_window (FlatpakInstance *instance)
+{
+  /* TODO: implement */
+  return FALSE;
+}
+
+static char *
+flatpak_instance_get_display_name (FlatpakInstance *instance)
+{
+  const char *app_id = flatpak_instance_get_app (instance);
+  if (app_id[0] != 0)
+    {
+      g_autofree char *desktop_id = NULL;
+      g_autoptr(GAppInfo) info = NULL;
+
+      desktop_id = g_strconcat (app_id, ".desktop", NULL);
+      info = (GAppInfo*)g_desktop_app_info_new (desktop_id);
```

Comment:
This "display name" is from the desktop file, and thus controlled by the app and is not trustworthy. Where is it displayed?

Reply:
This is going into a "Allow $APP to run in the background ?" notification. I've copied this approach from xdg-desktop-portal-gtk, where we already use this display name in various cases to talk to the user about an application request. 

Do you think we should add a ($APP_ID) behind it to make it trustworthy? After all, the display name from the desktop file is what we use to represent the app to the user in the shell, anyway

Reply:
Yeah, not clear what the best approach here is. On one hand, we can't ever get away from trusting the desktop file name, because that is what the user clicks on when launching apps. On the other hand, it makes it easy to fake it. It will get noticed pretty soon though...
Can we have some kind of details expander with the app id?